SQL Store.adoc

```diff
@@ -0,0 +1,251 @@
+
+[cols="1,2,1"]
+|===
+| *Query*
+| *Description*
+| *Example*
+
+| *SELECT ALL*
+| The query to use when bulk-loading data into the cache. Columns *MUST* be explicitly listed (wildcards are *NOT* allowed). Expressions are allowed but they *MUST* be labelled.
+| `select isbn, title from books`
+
+| *SELECT*
+| The query to use when loading a single entry into the cache. Columns *MUST* be explicitly listed (wildcards are *NOT* allowed). Expressions are allowed but they *MUST* be labelled.
+| `select isbn, title from books where isin = ?`
+
+| *UPSERT* 
+| The query to use when inserting/updating entrys into the database.
+| `insert into books (isbn, title) values() ON DUPLICATE KEY UPDATE`
+
+| *DELETE*
+| The query to use when a single entry is deleted.
+| `delete from books where id = ?`
+
+| *DELETE ALL*
+| The query to use when the cache is cleared.
+| `delete from books/truncate table books`
+
+|===
+
+The proposed XML configuration for this mode:
+
+[source,xml]
+----
+<local-cache name="default">
+   <persistence>
+      <query-jdbc-store>
+         <data-source jndi-url="..." />
+         <!-- the queries element is required, including the names of all the columns which make up the key -->
+         <queries key-columns="isbn">
+            <select-all>SELECT isbn, title FROM books</select-all>
+            <select>SELECT isbn, title FROM books WHERE isbn = :key</select>
+            <delete-all>DELETE FROM books</delete-all>
+            <delete>DELETE FROM books WHERE isbn = :key</delete>
+            <upsert>INSERT INTO books (isbn, title) VALUES (:key, :value) ON CONFLICT (isbn) DO UPDATE SET title = :value</upsert>
+         </queries>
+         <!-- the schema element is optional. Defaults will be used otherwise -->
+         <schema name="books.proto" key-name="books_id.proto" package="sql_store"/>
+      </query-jdbc-store>
+   </persistence>
+</local-cache>
+----
+
+==== Query parameters
+
+Query parameters are named using the following conventions:
+
+* primitive keys or values use the `:key` or `:value` names
+* composite keys or values use the `:key` or `:value` names followed by a `.` and the name of the field, e.g. `:value.description`
+
+== Storage
+
+=== ProtoBuf schema
+
+==== User-supplied schema
+
+It should be possible for the user to specify her own schema name, which should be present in the schema registry. 
+The names of the fields in the schema **MUST** match the column names of the table.
+
+==== Generated schema
+
+===== Naming
+
+The default naming of the schema will use the following convention: `*cache-name*-sql-store.proto`.
+
+=== Type mapping
+
+==== Primitive
+
+Primitive type mapping is used when the table/query only returns two columns. The following conventions are used:
+
+* A string will be used for column types which can be represented as such (char, varchar, text, numbers, boolean)
```

Comment:
I believe "string type mapping" would be a better name. The "primitive" name makes it sound like it would use the primitive types in protobuf (or the primitive wrapper types in Java).

Reply:
I've gone with the term "simple".